iroh-bytes/src/provider.rs

```diff
@@ -52,9 +66,44 @@ pub trait BaoMap: Clone + Send + Sync + 'static {
     /// This can also be used for a membership test by just checking if there
     /// is an entry. Creating an entry should be cheap, any expensive ops should
     /// be deferred to the creation of the actual readers.
+    ///
+    /// It is not guaranteed that the entry is complete. A BaoMapMut would return
+    /// here both complete and partial entries, so that you can share partial entries.
     fn get(&self, hash: &Hash) -> Option<Self::Entry>;
 }
 
+/// A partial entry
+pub trait BaoPartialMapEntry<D: BaoPartialMap>: BaoMapEntry<D> {
+    /// A future that resolves to an writeable outboard
+    fn outboard_mut(&self) -> BoxFuture<'_, io::Result<D::OutboardMut>>;
+    /// A future that resolves to a writer that can be used to write the data
+    fn data_writer(&self) -> BoxFuture<'_, io::Result<D::DataWriter>>;
+}
+
+/// A mutable bao map
+pub trait BaoPartialMap: BaoMap {
```

Comment:
Thanks. Not much left.

One thing I am not quite sure about is whether the get and get_partial fns should be async. Currently I think they can be sync, which means that at least the knowledge whether we have a hash or not must be quickly available in memory.

I can see scenarios where that would not be the case, e.g. you want to just rely on the file system for a membership test. But I think it is nice to be able to assume that this will be fast.

Reply:
I will leave them sync for now. We can always make them async later without much overhead, but not much point doing it before we have a use case.